File: Data Structures & Algorithms/binary-search/submission-1.java
// 精确找target模版
// public int search(int[] nums, int target) {
//     int l = 0, r = nums.length - 1;      // 闭区间 [l, r]
//     while (l <= r) {                      // 注意 <=
//         int m = l + (r - l) / 2;
//         if (nums[m] == target) return m;
//         else if (nums[m] < target) l = m + 1;   // 小了 → 往右
//         else r = m - 1;                          // 大了 → 往左
//     }
//     return -1;                            // 没找到
// }

class Solution {
    public int search(int[] nums, int target) {
        int l = 0, r = nums.length - 1;
        while (l <= r) {
            int m = l + (r - l) / 2;
            if (nums[m] == target) return m;
            else if (nums[m] < target) l = m + 1;
            else if (nums[m] > target) r = m - 1;
        }
        return -1; 
    }
}

File: Data Structures & Algorithms/three-integer-sum/submission-1.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = []
        for i in range(len(nums)-2):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            j = i + 1
            k = len(nums) - 1
            tar = -nums[i]
            while j < k:
                s = nums[j] + nums[k]
                if s == tar:
                    ans.append([nums[i],nums[j],nums[k]])
                    j += 1
                    k -= 1
                    while j < k and nums[j] == nums[j-1]:
                        j += 1
                elif s < tar:
                    j += 1
                elif s > tar:
                    k -= 1
        return ans